widgets/hwmontemp_linux.lua

```diff
@@ -39,16 +42,21 @@ return helpers.setasyncall{
             input = "temp1_input"
         end
 
-        spawn.easy_async_with_shell(
-            "grep " .. warg[1] .. " -wl /sys/class/hwmon/*/name",
-            function (stdout, stderr, exitreason, exitcode)
-                if exitreason == "exit" and exitcode == 0 then
-                    local f = io.open(stdout:gsub("name%s+", input), "r")
-                    callback{ tonumber(f:read"*line") / 1000 }
-                    f:close()
-                else
-                    callback{}
-                end
-            end)
+        if not pathcache[warg[1]] then
+            spawn.easy_async_with_shell(
+                "grep " .. warg[1] .. " -wl /sys/class/hwmon/*/name",
+                function (stdout, stderr, exitreason, exitcode)
+                    if exitreason == "exit" and exitcode == 0 then
+                        pathcache[warg[1]] = stdout:gsub("name%s+", "")
+                    end
+                end)
+        end
+        if pathcache[warg[1]] then
```

Comment:
On the first run this is very likely to be reached before the `pathcache[warg[1]]` is assigned (also nit pick: please give `warg[1]` a name, it's used in quite a few places in this patch).  I imagine that you can call `callback` within `easy_async_with_shell` and in the `else`-clause of the block above.  You might need to define a function to avoid duplicating the file opening code though.

Reply:
Okay I seem to have misunderstood how this async stuff works, thanks for pointing out.

Well, as the entire worker function (to my understanding) is already asynchronous, would it be an option to replace the `spawn.easy_async_with_shell()` part with something synchronous that performs the `grep` call?

Regarding `warg[1]` you're absolutely right and it was pretty hard for me not to give it a name! But I had a look at other widgets and, together with _"However, declare a variable only when you need it"_ (from CONTRIBUTING.md), I got the impression that one shall really stick to the absolute minimum.

I'll come up with a new pull request including a name for `warg[1]` once we're done talking about the point above.

Reply:
The worker function (I assume you're referring to the `function (format, warg, callback)` one) is supposed to be asynchronous, which is made so by using `spawn.easy_async_with_shell()` while doing the heavy lifting.  When grep is not needed, I suppose that we can execute the code (which is just reading a tiny file) synchronously w/o risking making the WM unresponsive.

> However, declare a variable only when you need it

IIRC refers to not doing the variable declaration at the beginning of every function like in ancient C (after reading the surrounding sentences that's what I felt it meant).  For the name suggestion, according to [the docs you wrote](https://github.com/vicious-widgets/vicious#viciouswidgetshwmontemp), I think we can call it `sensor`.

Reply:
> Please don't, just push changes to this one directly
Shame on me, I didn't know you can actually work on pull requests and change things. I'll find out how to do so :)